frontend/deb/debroot.go

```diff
@@ -0,0 +1,438 @@
+		v := spec.Build.Env[k]
+		fmt.Fprintf(buf, "export %q=%q\n", k, v)
+	}
+
+	for _, step := range spec.Build.Steps {
+		fmt.Fprintln(buf)
+		fmt.Fprintln(buf, "(")
+
+		sorted := dalec.SortMapKeys(step.Env)
+		for _, k := range sorted {
+			v := step.Env[k]
+			fmt.Fprintf(buf, "	export %q=%q\n", k, v)
+		}
+
+		fmt.Fprintln(buf, step.Command)
+		fmt.Fprintln(buf, ")")
+	}
+
+	return buf.Bytes()
+}
+
+func createInstallScripts(worker llb.State, spec *dalec.Spec, dir string) []llb.State {
+	if spec.Artifacts.IsEmpty() {
+		return nil
+	}
+
+	states := make([]llb.State, 0, len(spec.Artifacts.Binaries)+len(spec.Artifacts.Manpages))
+	base := llb.Scratch().File(llb.Mkdir(dir, 0o755, llb.WithParents(true)))
+
+	installBuf := bytes.NewBuffer(nil)
+	writeInstallHeader := sync.OnceFunc(func() {
+		fmt.Fprintln(installBuf, "#!/usr/bin/dh-exec")
+		fmt.Fprintln(installBuf)
+	})
+
+	writeInstall := func(src, dir, name string) {
+		writeInstallHeader()
+
+		if filepath.Base(src) != name {
+			fmt.Fprintln(installBuf, src, "=>", filepath.Join(dir, name))
+			return
+		}
+
+		fmt.Fprintln(installBuf, src, dir+"/")
+	}
+
+	if len(spec.Artifacts.Binaries) > 0 {
+		sorted := dalec.SortMapKeys(spec.Artifacts.Binaries)
+		for _, key := range sorted {
+			cfg := spec.Artifacts.Binaries[key]
+			writeInstall(key, filepath.Join("/usr/bin", cfg.SubPath), cfg.ResolveName(key))
+		}
+	}
+
+	if len(spec.Artifacts.ConfigFiles) > 0 {
+		buf := bytes.NewBuffer(nil)
+		sorted := dalec.SortMapKeys(spec.Artifacts.ConfigFiles)
+		for _, p := range sorted {
+			cfg := spec.Artifacts.ConfigFiles[p]
+
+			dir := filepath.Join("/etc", cfg.SubPath)
+			name := cfg.ResolveName(p)
+			writeInstall(p, dir, name)
+			fmt.Fprintln(buf, filepath.Join(dir, name))
+		}
+
+		// See: https://man7.org/linux/man-pages/man5/deb-conffiles.5.html for tracking config files in packages
+		states = append(states, base.File(llb.Mkfile(filepath.Join(dir, "conffiles"), 0o640, buf.Bytes())))
+	}
+
+	if len(spec.Artifacts.Manpages) > 0 {
+		buf := bytes.NewBuffer(nil)
+
+		sorted := dalec.SortMapKeys(spec.Artifacts.Manpages)
+		for _, key := range sorted {
+			cfg := spec.Artifacts.Manpages[key]
+			if cfg.Name != "" || (cfg.SubPath != "" && cfg.SubPath != filepath.Base(filepath.Dir(key))) {
+				resolved := cfg.ResolveName(key)
+				writeInstall(key, filepath.Join("/usr/share/doc/manpages", spec.Name, cfg.SubPath), resolved)
+				continue
+			}
+			fmt.Fprintln(buf, key)
+		}
+		if buf.Len() > 0 {
+			states = append(states, base.File(llb.Mkfile(filepath.Join(dir, spec.Name+".manpages"), 0o640, buf.Bytes())))
+		}
+
+	}
+
+	if spec.Artifacts.Directories != nil {
+		buf := bytes.NewBuffer(nil)
+
+		sorted := dalec.SortMapKeys(spec.Artifacts.Directories.Config)
+		for _, name := range sorted {
+			fmt.Fprintln(buf, filepath.Join("/etc", name))
+		}
+
+		sorted = dalec.SortMapKeys(spec.Artifacts.Directories.State)
+		for _, name := range sorted {
+			fmt.Fprintln(buf, filepath.Join("/var/lib", name))
```

Comment:
Should these paths be hardcoded? Is `/var/lib` the guaranteed convention?

Reply:
Same as https://github.com/Azure/dalec/pull/87#discussion_r1725751256